{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7c3ad7de-c1eb-434f-b5fd-8245683b902f",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(RCy3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e072fc31-5836-4f8c-be5a-b830ab21472c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "var Channel = \"b37dbd11-6852-46c6-9ef0-158dc5e1b75d\"; \n",
       "\n",
       " var JupyterBridge = \"https://jupyter-bridge.cytoscape.org\"; \n",
       "\n",
       "  /*\n",
       "    These functions serve as a connector between a remote Jupyter server and Cytoscape.\n",
       "    They run in the user's browser, which also shows the Jupyter Notebook.\n",
       "\n",
       "    A remote Jupyter Notebook call to the py4cytoscape package is forwarded to the Jupyter Bridge,\n",
       "    which is a standalone server. The functions in this connector execute in the Jupyter Notebook\n",
       "    browser, which executes on the same PC as Cytoscape. So, that's 4 components: (A) remote\n",
       "    Jupyter Notebook, (B) separate Jupyter Bridge server, (C) this browser-based component, and\n",
       "    (D) Cytoscape. (A) is on a remote server, (B) is on a different remote server, and (C) and (D)\n",
       "    are on the user's PC.\n",
       "\n",
       "    (A) calls its py4cytoscape module, which forwards the request (in a JSON wrapper) to (B).\n",
       "    (C) picks up the request from (B), unpacks the request and forwards it to (D). (C) awaits a\n",
       "    reply from (D), and when it gets it, it forwards the reply (in a JSON wrapper) to (B).\n",
       "    (A)'s py4cytoscape module picks up the reply on (B) when it becomes available, unpacks it,\n",
       "    and returns it to (A).\n",
       "\n",
       "    A Jupyter Notebook can talk to only one Cytoscape (i.e., the one on the machine running the\n",
       "    Jupyter Notebook browser), and Cytoscape should be called by only one Jupyter Notebook. The\n",
       "    Jupyter Bridge differentiates between Notebook-Cytoscape conversations via a channel UUID.\n",
       "    The UUID is prepended to this browser component by py4Cytoscape, and the component is\n",
       "    started by the Jupyter Notebook. (I wish py4Cytoscape could start the component, too, but I\n",
       "    haven't figured out how to do that yet, so startup code *is* required in the Jupyter\n",
       "    Notebook.)\n",
       "\n",
       "    Note that for the case of a Jupyter server running on the same machine as Cytoscape, this\n",
       "    bridge isn't necessary because the Jupyter server's HTTP calls can easily connect to\n",
       "    Cytoscape over a localhost socket. So, the combination of Jupyter Bridge and this browser\n",
       "    component solves the problem of a Jupyter server (e.g., Google's Colab) that can't\n",
       "    connect to Cytoscape that sits behind a firewall.\n",
       "\n",
       "    The request represents an HTTP call that py4cytoscape would normally make via HTTP directly\n",
       "    to Cytoscape via localhost when both py4cytoscape and Cytoscape are running on the same machine.\n",
       " */\n",
       "\n",
       "const VERSION = '0.0.2'\n",
       "\n",
       "var showDebug; // Flag indicating whether to show Jupyter-bridge progress\n",
       "if (typeof showDebug === 'undefined') {\n",
       "    showDebug = false\n",
       "}\n",
       "if (showDebug) {\n",
       "    alert(\"Starting Jupyter-bridge browser component\")\n",
       "}\n",
       "\n",
       "//const JupyterBridge = 'http://127.0.0.1:5000' // for testing against local Jupyter-bridge\n",
       "var JupyterBridge; // URL of Jupyter-bridge server could be defined by assignment pre-pended to this file\n",
       "if (typeof JupyterBridge === 'undefined') {\n",
       "    JupyterBridge = 'https://jupyter-bridge.cytoscape.org' // for production\n",
       "}\n",
       "var Channel; // Unique constant that could be defined by assignment pre-pended to this file\n",
       "if (typeof Channel === 'undefined') { // ... but if not assigned, use a debugging value\n",
       "    Channel = 1\n",
       "}\n",
       "\n",
       "\n",
       "var httpR = new XMLHttpRequest(); // for sending reply to Jupyter-bridge\n",
       "var httpRE = new XMLHttpRequest(); // for sending backup error reply to Jupyter-bridge\n",
       "var httpC = new XMLHttpRequest(); // for sending command to Cytoscape\n",
       "var httpJ = new XMLHttpRequest(); // for fetching request from Jupyter-bridge\n",
       "\n",
       "const HTTP_OK = 200\n",
       "const HTTP_SYS_ERR = 500\n",
       "const HTTP_TIMEOUT = 408\n",
       "const HTTP_TOO_MANY = 429\n",
       "\n",
       "\n",
       " /* This function is useful if we want to rewrite the incoming URL to resolve just to our local one.\n",
       "    Doing this stops the Jupyter component from abusing this client to call out to endpoints other\n",
       "    than local Cytoscape. On the other hand, it makes it hard to detect when the Jupyter component\n",
       "    has specified a genuinely bad URL and really should get an error result. For now, we'll execute\n",
       "    the Jupyter-supplied URL and return the result, whatever it may be.\n",
       "\n",
       "const LocalCytoscape = 'http://127.0.0.1:1234'\n",
       "\n",
       "function parseURL(url) {\n",
       "    var reURLInformation = new RegExp([\n",
       "        '^(https?:)//', // protocol\n",
       "        '(([^:/?#]*)(?::([0-9]+))?)', // host (hostname and port)\n",
       "        '(/{0,1}[^?#]*)', // pathname\n",
       "        '(\\\\?[^#]*|)', // search\n",
       "        '(#.*|)$' // hash\n",
       "    ].join(''));\n",
       "    var match = url.match(reURLInformation);\n",
       "    return match && {\n",
       "        url: url,\n",
       "        protocol: match[1],\n",
       "        host: match[2],\n",
       "        hostname: match[3],\n",
       "        port: match[4],\n",
       "        pathname: match[5],\n",
       "        search: match[6],\n",
       "        hash: match[7]\n",
       "    }\n",
       "}\n",
       "*/\n",
       "\n",
       "function replyCytoscape(replyStatus, replyStatusText, replyText) {\n",
       "\n",
       "    // Clean up after Jupyter bridge accepts reply\n",
       "    httpR.onreadystatechange = function() {\n",
       "        if (httpR.readyState === 4) {\n",
       "            if (showDebug) {\n",
       "                console.log(' status from queue_reply: ' + httpR.status + ', reply: ' + httpR.responseText)\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "\n",
       "    httpR.onerror = function() {\n",
       "        // Clean up after Jupyter bridge accepts backup reply\n",
       "        httpRE.onreadystatechange = function() {\n",
       "            if (httpRE.readyState === 4) {\n",
       "                if (showDebug) {\n",
       "                    console.log(' status from backup queue_reply: ' + httpRE.status + ', reply: ' + httpRE.responseText)\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "\n",
       "        if (showDebug) {\n",
       "            console.log(' error from queue_reply -- could be Jupyter-Bridge server reject')\n",
       "        }\n",
       "        var errReply = {'status': HTTP_SYS_ERR, 'reason': 'Jupyter-Bridge rejected reply', 'text': 'Possibly reply is too long for Jupyter-Bridge server'}\n",
       "        httpRE.open('POST', jupyterBridgeURL, true)\n",
       "        httpRE.setRequestHeader('Content-Type', 'text/plain')\n",
       "        httpRE.send(JSON.stringify(errReply))\n",
       "    }\n",
       "\n",
       "    var reply = {'status': replyStatus, 'reason': replyStatusText, 'text': replyText}\n",
       "\n",
       "    // Send reply to Jupyter bridge\n",
       "    var jupyterBridgeURL = JupyterBridge + '/queue_reply?channel=' + Channel\n",
       "    if (showDebug) {\n",
       "        console.log('Starting queue to Jupyter bridge: ' + jupyterBridgeURL)\n",
       "    }\n",
       "    httpR.open('POST', jupyterBridgeURL, true)\n",
       "    httpR.setRequestHeader('Content-Type', 'text/plain')\n",
       "    httpR.send(JSON.stringify(reply))\n",
       "}\n",
       "\n",
       "function callCytoscape(callSpec) {\n",
       "\n",
       "    // Captures Cytoscape reply and sends it on\n",
       "    httpC.onreadystatechange = function() {\n",
       "        if (httpC.readyState === 4) {\n",
       "            if (showDebug) {\n",
       "                console.log(' status from CyREST: ' + httpC.status + ', statusText: ' + httpC.statusText + ', reply: ' + httpC.responseText)\n",
       "            }\n",
       "            // Note that httpC.status is 0 if the URL can't be reached *OR* there is a CORS violation.\n",
       "            // I wish I could tell the difference because for a CORS violation, I'd return a 404,\n",
       "            // which would roughly match what Python's native request package would return.\n",
       "            // The practical consequence is that the ultimate caller (e.g., py4cytoscape)\n",
       "            // returns different exceptions, depending on wither this module is doing the\n",
       "            // HTTP operation or the native Python requests package is. This is minor, but\n",
       "            // messes up tests that verify the exception type.\n",
       "            replyCytoscape(httpC.status, httpC.statusText, httpC.responseText)\n",
       "            waitOnJupyterBridge()\n",
       "        }\n",
       "    }\n",
       "\n",
       "//  Build up request to Cytoscape, making sure host is local.\n",
       "//    Too heavy handed: localURL = LocalCytoscape + parseURL(callSpec.url).pathname\n",
       "    var localURL = callSpec.url // Try using what was passed in ... is there a security risk??\n",
       "\n",
       "    if (showDebug) {\n",
       "        console.log('Command to CyREST: ' + callSpec.command + ' (' + localURL + ')')\n",
       "        if (callSpec.params) {\n",
       "            console.log(' params: ' + JSON.stringify(callSpec.params))\n",
       "        }\n",
       "        if (callSpec.headers) {\n",
       "            console.log(' header: ' + JSON.stringify(callSpec.headers))\n",
       "        }\n",
       "        if (callSpec.data) {\n",
       "            console.log('   data: ' + JSON.stringify(callSpec.data))\n",
       "        }\n",
       "    }\n",
       "\n",
       "    if (callSpec.command === 'webbrowser') {\n",
       "        if (window.open(callSpec.url)) {\n",
       "            replyCytoscape(HTTP_OK, 'OK', '')\n",
       "        } else {\n",
       "            replyCytoscape(HTTP_SYS_ERR, 'BAD BROWSER OPEN', '')\n",
       "        }\n",
       "        waitOnJupyterBridge()\n",
       "    } else if (callSpec.command === 'version') {\n",
       "        replyCytoscape(HTTP_OK, 'OK',\n",
       "            JSON.stringify({\"jupyterBridgeVersion\": VERSION}))\n",
       "        waitOnJupyterBridge()\n",
       "    } else {\n",
       "        var joiner = '?'\n",
       "        for (let param in callSpec.params) {\n",
       "            localURL = localURL + joiner + param + '=' + encodeURIComponent(callSpec.params[param])\n",
       "            joiner = '&'\n",
       "        }\n",
       "\n",
       "        httpC.open(callSpec.command, localURL, true)\n",
       "        for (let header in callSpec.headers) {\n",
       "            httpC.setRequestHeader(header, callSpec.headers[header])\n",
       "        }\n",
       "\n",
       "        // Send request to Cytoscape ... reply goes to onreadystatechange handler\n",
       "        httpC.send(JSON.stringify(callSpec.data))\n",
       "    }\n",
       "}\n",
       "\n",
       "function waitOnJupyterBridge() {\n",
       "\n",
       "    // Captures request from Jupyter bridge\n",
       "    httpJ.onreadystatechange = function() {\n",
       "        if (httpJ.readyState === 4) {\n",
       "            if (showDebug) {\n",
       "                console.log(' status from dequeue_request: ' + httpJ.status + ', reply: ' + httpJ.responseText)\n",
       "            }\n",
       "            try {\n",
       "                if (httpJ.status == HTTP_TOO_MANY) {\n",
       "                    // Nothing more to do ... the browser has created too many listeners,\n",
       "                    // and it's time to stop listening because the server saw a listener\n",
       "                    // listening on this channel before we got there.\n",
       "                    console.log('  shutting down because of redundant reader on channel: ' + Channel)\n",
       "                } else {\n",
       "                    if (httpJ.status === HTTP_TIMEOUT) {\n",
       "                        waitOnJupyterBridge()\n",
       "                    } else {\n",
       "                        callCytoscape(JSON.parse(httpJ.responseText))\n",
       "                    }\n",
       "                }\n",
       "            } catch(err) {\n",
       "                if (showDebug) {\n",
       "                    console.log(' exception calling Cytoscape: ' + err)\n",
       "                }\n",
       "                // Bad responseText means something bad happened that we don't understand.\n",
       "                // Go wait on another request, as there's nothing to call Cytoscape with.\n",
       "                waitOnJupyterBridge()\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "\n",
       "    // Wait for request from Jupyter bridge\n",
       "    var jupyterBridgeURL = JupyterBridge + '/dequeue_request?channel=' + Channel\n",
       "    if (showDebug) {\n",
       "        console.log('Starting dequeue on Jupyter bridge: ' + jupyterBridgeURL)\n",
       "    }\n",
       "    httpJ.open('GET', jupyterBridgeURL, true)\n",
       "    httpJ.send()\n",
       "}\n",
       "\n",
       "// This kicks off a loop that ends by calling waitOnJupyterBridge again. This first call\n",
       "// ejects any dead readers before we start a read\n",
       "waitOnJupyterBridge() // Wait for message from Jupyter bridge, execute it, and return reply\n",
       "\n",
       "if (showDebug) {\n",
       "    alert(\"Jupyter-bridge browser component is started on \" + JupyterBridge + ', channel ' + Channel)\n",
       "}\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "browserClientJs <- getBrowserClientJs()\n",
    "IRdisplay::display_javascript(browserClientJs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2566d0ab-994e-4502-820d-0338d7677a7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "You are connected to Cytoscape!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cytoscapePing()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40567949-b17a-4fb4-a822-3529ac5fe536",
   "metadata": {},
   "source": [
    "# Cytoscape automation with RCy3\n",
    "\n",
    "Authors:\n",
    "- Kozo Nishida [RIKEN],\n",
    "- Alex Pico [Gladstone Institute].\n",
    "\n",
    "Last modified: 22 Oct, 2021.\n",
    "\n",
    "## Overview\n",
    "\n",
    "### Description\n",
    "\n",
    "Along with the topic of your workshop, include how students can expect\n",
    "to spend their time. For the description may also include information\n",
    "about what type of workshop it is (e.g. instructor-led live demo, lab,\n",
    "lecture + lab, etc.). Instructors are strongly recommended to provide\n",
    "completely worked examples for lab sessions, and a set of stand-alone\n",
    "notes that can be read and understood outside of the workshop.\n",
    "\n",
    "### Pre-requisites\n",
    "\n",
    "List any workshop prerequisites, for example:\n",
    "\n",
    "* Basic knowledge of R syntax\n",
    "* Familiarity with the GenomicRanges class\n",
    "* Familiarity with xyz vignette (provide link)\n",
    "\n",
    "List relevant background reading for the workshop, including any\n",
    "theoretical background you expect students to have.\n",
    "\n",
    "* List any textbooks, papers, or other reading that students should be\n",
    "  familiar with. Include direct links where possible.\n",
    "\n",
    "### Participation\n",
    "\n",
    "Describe how students will be expected to participate in the workshop.\n",
    "\n",
    "### _R_ / _Bioconductor_ packages used\n",
    "\n",
    "List any _R_ / _Bioconductor_ packages that will be explicitly covered.\n",
    "\n",
    "### Time outline\n",
    "\n",
    "An example for a 45-minute workshop:\n",
    "\n",
    "| Activity                     | Time |\n",
    "|------------------------------|------|\n",
    "| Cytoscape Automation         | 15m  |\n",
    "| Automation Use Cases         | 20m  |\n",
    "| Wrap-up and \"Exam\"           | 5m   |\n",
    "\n",
    "### Workshop goals and objectives\n",
    "\n",
    "By the end of this workshop you should be able to: \n",
    "- Command programmatic control over Cytoscape\n",
    "- Integrate Cytoscape into your bioinformatics pipelines\n",
    "- Publish, share and export networks online\n",
    "\n",
    "### Learning goals\n",
    "\n",
    "Some examples:\n",
    "\n",
    "* describe how to...\n",
    "* identify methods for...\n",
    "* understand the difference between...\n",
    "\n",
    "### Learning objectives\n",
    "\n",
    "* analyze xyz data to produce...\n",
    "* create xyz plots\n",
    "* evaluate xyz data for artifacts\n",
    "\n",
    "## Workshop\n",
    "\n",
    "Divide the workshop into sections (`## A Section`). Include\n",
    "fully-evaluated _R_ code chunks. Develop exercises and solutions, and\n",
    "anticipate that your audience will walk through the code with you, or\n",
    "work on the code idependently -- do not be too ambitious in the\n",
    "material that you present.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c18f83cd-a155-4dfd-8d98-18e4a6ca67eb",
   "metadata": {},
   "source": [
    "## Getting started\n",
    "Confirm that Cytoscape is installed and opened"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9a12e4f5-b42c-4cfc-9a5e-141f74b47af3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>apiVersion</dt><dd>'v1'</dd><dt>cytoscapeVersion</dt><dd>'3.9.0'</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[apiVersion] 'v1'\n",
       "\\item[cytoscapeVersion] '3.9.0'\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "apiVersion\n",
       ":   'v1'cytoscapeVersion\n",
       ":   '3.9.0'\n",
       "\n"
      ],
      "text/plain": [
       "      apiVersion cytoscapeVersion \n",
       "            \"v1\"          \"3.9.0\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cytoscapeVersionInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a049633-ec4b-4b54-9d27-fb45b527aec5",
   "metadata": {},
   "source": [
    "### Browse available functions, commands and arguments\n",
    "Depending on what apps you have installed there is different functionality available.\n",
    "\n",
    "To see all the functions available in RCy3 package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e1df4f77-8373-4935-8723-7afb93234cdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Documentation for package ‘RCy3’\n",
       "\n",
       "\n",
       "\t\tInformation on package ‘RCy3’\n",
       "\n",
       "Description:\n",
       "\n",
       "Package:              RCy3\n",
       "Type:                 Package\n",
       "Title:                Functions to Access and Control Cytoscape\n",
       "Version:              2.13.9\n",
       "Date:                 2021-10-11\n",
       "Authors@R:            c(person(\"Alex\", \"Pico\", role = c(\"aut\",\"cre\"),\n",
       "                      email = \"alex.pico@gladstone.ucsf.edu\", comment =\n",
       "                      c(ORCID = \"0000-0001-5706-2163\")),\n",
       "                      person(\"Tanja\", \"Muetze\", role = \"aut\", email =\n",
       "                      \"tanja.muetze14@alumni.imperial.ac.uk\"),\n",
       "                      person(\"Paul\", \"Shannon\", role = \"aut\", email =\n",
       "                      \"pshannon@systemsbiology.org\"), person(\"Ruth\",\n",
       "                      \"Isserlin\", role = \"ctb\"), person(\"Shraddha\",\n",
       "                      \"Pai\", role = \"ctb\"), person(\"Julia\",\n",
       "                      \"Gustavsen\", role = \"ctb\"), person(\"Georgi\",\n",
       "                      \"Kolishovski\", role = \"ctb\"), person(\"Yihang\",\n",
       "                      \"Xin\", role = \"ctb\") )\n",
       "Imports:              httr, methods, RJSONIO, XML, utils, BiocGenerics,\n",
       "                      stats, graph, fs, uuid, uchardet, glue, RCurl,\n",
       "                      base64url, base64enc, IRkernel, IRdisplay,\n",
       "                      RColorBrewer\n",
       "Suggests:             BiocStyle, knitr, rmarkdown, igraph, grDevices\n",
       "SystemRequirements:   Cytoscape (>= 3.7.1), CyREST (>= 3.8.0)\n",
       "Description:          Vizualize, analyze and explore networks using\n",
       "                      Cytoscape via R. Anything you can do using the\n",
       "                      graphical user interface of Cytoscape, you can\n",
       "                      now do with a single RCy3 function.\n",
       "License:              MIT + file LICENSE\n",
       "URL:                  https://github.com/cytoscape/RCy3\n",
       "BugReports:           https://github.com/cytoscape/RCy3/issues\n",
       "LazyLoad:             yes\n",
       "biocViews:            Visualization, GraphAndNetwork, ThirdPartyClient,\n",
       "                      Network\n",
       "NeedsCompilation:     no\n",
       "VignetteBuilder:      knitr\n",
       "RoxygenNote:          7.1.2\n",
       "RemoteType:           github\n",
       "RemoteHost:           api.github.com\n",
       "RemoteRepo:           RCy3\n",
       "RemoteUsername:       cytoscape\n",
       "RemoteRef:            HEAD\n",
       "RemoteSha:            780e6d520391c22a086c910d641c262871b0a3b0\n",
       "GithubRepo:           RCy3\n",
       "GithubUsername:       cytoscape\n",
       "GithubRef:            HEAD\n",
       "GithubSHA1:           780e6d520391c22a086c910d641c262871b0a3b0\n",
       "Packaged:             2021-10-22 12:08:35 UTC; root\n",
       "Author:               Alex Pico [aut, cre]\n",
       "                      (<https://orcid.org/0000-0001-5706-2163>), Tanja\n",
       "                      Muetze [aut], Paul Shannon [aut], Ruth Isserlin\n",
       "                      [ctb], Shraddha Pai [ctb], Julia Gustavsen [ctb],\n",
       "                      Georgi Kolishovski [ctb], Yihang Xin [ctb]\n",
       "Maintainer:           Alex Pico <alex.pico@gladstone.ucsf.edu>\n",
       "Built:                R 4.1.1; ; 2021-10-22 12:08:37 UTC; unix\n",
       "\n",
       "Index:\n",
       "\n",
       ".exportShowImage        .exportShowImage\n",
       ".getDefaultSandbox      .getDefaultSandbox\n",
       ".getRequester           .getRequester\n",
       "addAnnotationBoundedText\n",
       "                        Add Bounded Text Annotation\n",
       "addAnnotationImage      Add Image Annotation\n",
       "addAnnotationShape      Add Shape Annotation\n",
       "addAnnotationText       Add Text Annotation\n",
       "addCyEdges              Add CyEdges\n",
       "addCyNodes              Add CyNodes\n",
       "AddToGroup              Add to Group\n",
       "analyzeNetwork          Analyze Network\n",
       "applyFilter             Apply Filter\n",
       "bundleEdges             Bundle Edges\n",
       "checkNotebookIsRunning\n",
       "                        checkNotebookIsRunning\n",
       "checkRunningRemote      checkRunningRemote\n",
       "clearEdgeBends          Clear Edge Bends\n",
       "clearEdgePropertyBypass\n",
       "                        Clear Edge Property Bypass\n",
       "clearNetworkCenterBypass\n",
       "                        Clear Network Center Bypass\n",
       "clearNetworkPropertyBypass\n",
       "                        Clear Network Property Bypass\n",
       "clearNetworkZoomBypass\n",
       "                        Clear Network Zoom Bypass\n",
       "clearNodeOpacityBypass\n",
       "                        Clear Node Opacity Bypass\n",
       "clearNodePropertyBypass\n",
       "                        Clear Node Property Bypass\n",
       "clearSelection          Clear Selection\n",
       "cloneNetwork            Clone a Cytoscape Network\n",
       "closeSession            Close Session\n",
       "collapseGroup           Collapse Group\n",
       "commandEcho             Command Echo\n",
       "commandOpenDialog       Command Open Dialog\n",
       "commandPause            Command Pause\n",
       "commandQuit             Command Quit\n",
       "commandRunFile          Command Run File\n",
       "commandsAPI             Open Swagger docs for CyREST Commands API\n",
       "commandsGET             Commands GET\n",
       "commandsHelp            Commands Help\n",
       "commandSleep            Command Sleep\n",
       "commandsPOST            Commands POST\n",
       "commandsRun             Run a Command\n",
       "copyVisualStyle         Copy Visual Style\n",
       "createColumnFilter      Create Column Filter\n",
       "createCompositeFilter   Create Composite Filter\n",
       "createDegreeFilter      Create Degree Filter\n",
       "createGraphFromNetwork\n",
       "                        createGraphFromNetwork\n",
       "createGroup             Create Group\n",
       "createGroupByColumn     Create Group by Column\n",
       "createIgraphFromNetwork\n",
       "                        Create an igraph network from a Cytoscape\n",
       "                        network\n",
       "createNetworkFromDataFrames\n",
       "                        Create a network from data frames\n",
       "createNetworkFromGraph\n",
       "                        Create Network From Graph\n",
       "createNetworkFromIgraph\n",
       "                        Create a Cytoscape network from an igraph\n",
       "                        network\n",
       "createSubnetwork        Create subnetwork from existing network\n",
       "createVisualStyle       Create a visual style from components\n",
       "cybrowserClose          Cybrowser Close\n",
       "cybrowserDialog         Cybrowser Dialog\n",
       "cybrowserHide           Cybrowser Hide\n",
       "cybrowserList           Cybrowser List\n",
       "cybrowserSend           Cybrowser Send\n",
       "cybrowserShow           Cybrowser Show\n",
       "cybrowserVersion        Cybrowser Version\n",
       "cyrestAPI               Open Swagger docs for CyREST API\n",
       "cyrestDELETE            CyREST DELETE\n",
       "cyrestGET               CyREST GET\n",
       "cyrestPOST              CyREST POST\n",
       "cyrestPUT               CyREST PUT\n",
       "cytoscapeApiVersions    Available CyREST API Versions\n",
       "cytoscapeFreeMemory     Free Up Unused Memory for Cytoscape\n",
       "cytoscapeMemoryStatus   Memory Available to Cytoscape\n",
       "cytoscapeNumberOfCores\n",
       "                        Number of Processors Available to Cytoscape\n",
       "cytoscapePing           Ping Cytoscape\n",
       "cytoscapeVersionInfo    Cytoscape and CyREST API Versions\n",
       "deleteAllNetworks       Delete All Networks\n",
       "deleteAnnotation        Delete Annotation\n",
       "deleteDuplicateEdges    Delete Duplicate Edges\n",
       "deleteGroup             Delete (or Ungroup) a Group\n",
       "deleteNetwork           Delete Network\n",
       "deleteSelectedEdges     Delete Selected Edges\n",
       "deleteSelectedNodes     Delete Selected Nodes\n",
       "deleteSelfLoops         Delete Self Loops\n",
       "deleteStyleMapping      Delete Style Mapping\n",
       "deleteTableColumn       Delete a table column\n",
       "deleteVisualStyle       Delete Visual Style\n",
       "diffusionAdvanced       Diffusion Advanced\n",
       "diffusionBasic          Diffusion Basic\n",
       "disableApp              Disable App\n",
       "dockPanel               Dock Panel\n",
       "doInitializeSandbox     doInitializeSandbox\n",
       "doRequestRemote         doRequestRemote\n",
       "doSetSandbox            doSetSandbox\n",
       "enableApp               Enable App\n",
       "expandGroup             Expand Group\n",
       "exportFilters           Export Filters\n",
       "exportImage             Export Image\n",
       "exportNetwork           Export Network\n",
       "exportNetworkToNDEx     Export Network To NDEx\n",
       "exportVisualStyles      Export Visual Styles\n",
       "findRemoteCytoscape     findRemoteCytoscape\n",
       "fitContent              Fit Content\n",
       "floatPanel              Float Panel\n",
       "getAbsSandboxPath       getAbsSandboxPath\n",
       "getAllEdges             Get All Edges\n",
       "getAllNodes             Get All Nodes\n",
       "getAllStyleMappings     Get All Style Mappings\n",
       "getAnnotationList       Get Annotation List\n",
       "getAppInformation       Get App Information\n",
       "getAppStatus            App Status\n",
       "getAppUpdates           List Apps With Updates\n",
       "getArrowShapes          Get Arrow Shapes\n",
       "getAvailableApps        List Available Apps\n",
       "getBackgroundColorDefault\n",
       "                        Get Background Color Default\n",
       "getBrowserClientChannel\n",
       "                        getBrowserClientChannel\n",
       "getBrowserClientJs      getBrowserClientJs\n",
       "getCollectionList       Get Collection List\n",
       "getCollectionName       Get Collection Name\n",
       "getCollectionNetworks   Get Collection Networks\n",
       "getCollectionSuid       Get Collection Suid\n",
       "getCurrentSandbox       getCurrentSandbox\n",
       "getCurrentSandboxName   getCurrentSandboxName\n",
       "getCurrentSandboxPath   getCurrentSandboxPath\n",
       "getCurrentStyle         Get Current Style\n",
       "getDefaultSandbox       getDefaultSandbox\n",
       "getDefaultSandboxPath   getDefaultSandboxPath\n",
       "getDisabledApps         List Disabled Apps\n",
       "getEdgeColor            Get Edge Color\n",
       "getEdgeCount            Get Edge Count\n",
       "getEdgeInfo             Get Edge Information\n",
       "getEdgeLineStyle        Get Edge Line Style\n",
       "getEdgeLineWidth        Get Edge Line Width\n",
       "getEdgeProperty         Get Edge Property Values\n",
       "getEdgeSelectionColorDefault\n",
       "                        Get Edge Selection Color Default\n",
       "getEdgeTargetArrowShape\n",
       "                        Get Edge Target Arrow Shape\n",
       "getFilterList           Get Filter List\n",
       "getFirstNeighbors       Get list of nodes neighboring provided list\n",
       "getGroupInfo            Get Group Information\n",
       "getInstalledApps        List Installed Apps\n",
       "getJupyterBridgeURL     getJupyterBridgeURL\n",
       "getLayoutNameMapping    Get Layout Name Mapping\n",
       "getLayoutNames          Get Layout Names\n",
       "getLayoutPropertyNames\n",
       "                        Get Layout Property Names\n",
       "getLayoutPropertyType   Get Layout Property Type\n",
       "getLayoutPropertyValue\n",
       "                        Get Layout Property Value\n",
       "getLineStyles           Get Line Styles\n",
       "getNetworkCenter        Get Network Center\n",
       "getNetworkCount         Get the number of Cytoscape networks\n",
       "getNetworkList          Get the list of Cytoscape networks\n",
       "getNetworkName          Get the name of a network\n",
       "getNetworkNDExId        Get Network NDEx Id\n",
       "getNetworkProperty      Get Network Property Values\n",
       "getNetworkSuid          Get the SUID of a network\n",
       "getNetworkViews         Get Network Views\n",
       "getNetworkViewSuid      Get the SUID of a network view\n",
       "getNetworkZoom          Get Network Zoom\n",
       "getNodeColor            Get Node Color\n",
       "getNodeCount            Get Node Count\n",
       "getNodeHeight           Get Node Height\n",
       "getNodePosition         Get Node Position\n",
       "getNodeProperty         Get Node Property Values\n",
       "getNodeSelectionColorDefault\n",
       "                        Get Node Selection Color Default\n",
       "getNodeShapes           Get Node Shapes\n",
       "getNodeSize             Get Node Size\n",
       "getNodeWidth            Get Node Width\n",
       "getNotebookIsRunning    getNotebookIsRunning\n",
       "getSandboxReinitialize\n",
       "                        getSandboxReinitialize\n",
       "getSelectedEdgeCount    Get Selected Edge Count\n",
       "getSelectedEdges        Get Selected Edges\n",
       "getSelectedNodeCount    Get Selected Node Count\n",
       "getSelectedNodes        Get Selected Nodes\n",
       "getStyleDependencies    Get the values of dependencies in a style\n",
       "getStyleMapping         Get Style Mapping\n",
       "getTableColumnNames     Get Table Column Names\n",
       "getTableColumns         Get table column values\n",
       "getTableColumnTypes     Get Table Column Types\n",
       "getTableValue           Get table cell value\n",
       "getUninstalledApps      List Uninstalled Apps\n",
       "getVisualPropertyDefault\n",
       "                        Get Visual Property Default\n",
       "getVisualPropertyNames\n",
       "                        Get Visual Property Names\n",
       "getVisualStyleNames     Get Visual Style Names\n",
       "groupAnnotation         Group Annotation\n",
       "hideAllPanels           Hide All Panels\n",
       "hideEdges               Hide Edges\n",
       "hideNodes               Hide Nodes\n",
       "hidePanel               Hide Panel\n",
       "hideSelectedEdges       Hide Selected Edges\n",
       "hideSelectedNodes       Hide Selected Nodes\n",
       "importFilters           Import Filters\n",
       "importNetworkFromFile   Import Network From File\n",
       "importNetworkFromNDEx   Import Network From NDEx\n",
       "importVisualStyles      Import Visual Styles\n",
       "installApp              Install App\n",
       "invertEdgeSelection     Invert Edge Selection\n",
       "invertNodeSelection     Invert Node Selection\n",
       "layoutCopycat           Copy a layout from one network to another\n",
       "layoutNetwork           Apply a layout to a network\n",
       "listGroups              List Groups\n",
       "loadTableData           Loads data into Cytoscape tables keyed by row\n",
       "lockNodeDimensions      Lock Node Dimensions\n",
       "makeSimpleGraph         Make Simple Graph\n",
       "makeSimpleIgraph        Make Simple Igraph\n",
       "mapTableColumn          Map Table Column\n",
       "mapVisualProperty       Creates a mapping between an attribute and a\n",
       "                        visual property\n",
       "matchArrowColorToEdge   Match Arrow Color To Edge\n",
       "mergeNetworks           Merge Networks\n",
       "notebookExportShowImage\n",
       "                        notebookExportShowImage\n",
       "notebookShowImage       notebookShowImage\n",
       "openAppStore            Open App Store Page\n",
       "openSession             Open Session File or URL\n",
       "paletteColorBrewerAccent\n",
       "                        paletteColorBrewerAccent Qualitative\n",
       "paletteColorBrewerBlues\n",
       "                        paletteColorBrewerBlues Sequential\n",
       "paletteColorBrewerBrBG\n",
       "                        paletteColorBrewerBrBG Divergent\n",
       "paletteColorBrewerBuGn\n",
       "                        paletteColorBrewerBuGn Sequential\n",
       "paletteColorBrewerBuPu\n",
       "                        paletteColorBrewerBuPu Sequential\n",
       "paletteColorBrewerDark2\n",
       "                        paletteColorBrewerDark2 Qualitative\n",
       "paletteColorBrewerGnBu\n",
       "                        paletteColorBrewerGnBu Sequential\n",
       "paletteColorBrewerGreens\n",
       "                        paletteColorBrewerGreens Sequential\n",
       "paletteColorBrewerGreys\n",
       "                        paletteColorBrewerGreys Sequential\n",
       "paletteColorBrewerOranges\n",
       "                        paletteColorBrewerOranges Sequential\n",
       "paletteColorBrewerOrRd\n",
       "                        paletteColorBrewerOrRd Sequential\n",
       "paletteColorBrewerPaired\n",
       "                        paletteColorBrewerPaired Qualitative\n",
       "paletteColorBrewerPastel1\n",
       "                        paletteColorBrewerPastel1 Qualitative\n",
       "paletteColorBrewerPastel2\n",
       "                        paletteColorBrewerPastel2 Qualitative\n",
       "paletteColorBrewerPiYG\n",
       "                        paletteColorBrewerPiYG Divergent\n",
       "paletteColorBrewerPRGn\n",
       "                        paletteColorBrewerPRGn Divergent\n",
       "paletteColorBrewerPuBu\n",
       "                        paletteColorBrewerPuBu Sequential\n",
       "paletteColorBrewerPuBuGn\n",
       "                        paletteColorBrewerPuBuGn Sequential\n",
       "paletteColorBrewerPuOr\n",
       "                        paletteColorBrewerPuOr Divergent\n",
       "paletteColorBrewerPuRd\n",
       "                        paletteColorBrewerPuRd Sequential\n",
       "paletteColorBrewerPurples\n",
       "                        paletteColorBrewerPurples Sequential\n",
       "paletteColorBrewerRdBu\n",
       "                        paletteColorBrewerRdBu Divergent\n",
       "paletteColorBrewerRdPu\n",
       "                        paletteColorBrewerRdPu Sequential\n",
       "paletteColorBrewerRdYlBu\n",
       "                        paletteColorBrewerRdYlBu Divergent\n",
       "paletteColorBrewerReds\n",
       "                        paletteColorBrewerReds Sequential\n",
       "paletteColorBrewerSet1\n",
       "                        paletteColorBrewerSet1 Qualitative\n",
       "paletteColorBrewerSet2\n",
       "                        paletteColorBrewerSet2 Qualitative\n",
       "paletteColorBrewerSet3\n",
       "                        paletteColorBrewerSet3 Qualitative\n",
       "paletteColorBrewerYlGn\n",
       "                        paletteColorBrewerYlGn Sequential\n",
       "paletteColorBrewerYlGnBu\n",
       "                        paletteColorBrewerYlGnBu Sequential\n",
       "paletteColorBrewerYlOrBr\n",
       "                        paletteColorBrewerYlOrBr Sequential\n",
       "paletteColorBrewerYlOrRd\n",
       "                        paletteColorBrewerYlOrRd Sequential\n",
       "paletteColorRandom      paletteColorRandom Qualitative\n",
       "RCy3                    RCy3: Functions to Access and Control Cytoscape\n",
       "RemoveFromGroup         Remove from Group\n",
       "removeNodeCustomGraphics\n",
       "                        Remove Node Custom Graphics\n",
       "renameNetwork           Rename a network\n",
       "renameTableColumn       Rename Table Column\n",
       "resetDefaultSandbox     resetDefaultSandbox\n",
       "runningRemoteCheck      runningRmoteCheck\n",
       "sandboxGetFileInfo      sandboxGetFileInfo\n",
       "sandboxGetFrom          sandboxGetFrom\n",
       "sandboxInitializer      sandboxInitializer\n",
       "sandboxRemove           sandboxRemove\n",
       "sandboxRemoveFile       sandboxRemoveFile\n",
       "sandboxSendTo           sandboxSendTo\n",
       "sandboxSet              sandboxSet\n",
       "sandboxUrlTo            sandboxUrlTo\n",
       "saveSession             Save Session to File\n",
       "selectAllEdges          Select all edges\n",
       "selectAllNodes          Select all nodes\n",
       "selectEdges             Select Edges\n",
       "selectEdgesAdjacentToSelectedNodes\n",
       "                        Select Edges Adjacent To Selected Nodes\n",
       "selectEdgesConnectingSelectedNodes\n",
       "                        Select the edges connecting selected nodes in\n",
       "                        Cytoscape Network\n",
       "selectFirstNeighbors    Select first neighbor nodes\n",
       "selectNodes             Select Nodes\n",
       "selectNodesConnectedBySelectedEdges\n",
       "                        Select Nodes Connected By Selected Edges\n",
       "setBackgroundColorDefault\n",
       "                        Set Background Color Default\n",
       "setCatchupFilterSecs    Set Catchup Filter Delay\n",
       "setCatchupNetworkSecs   Set Catchup Network Delay\n",
       "setCurrentNetwork       Set current network\n",
       "setCurrentSandbox       setCurrentSandbox\n",
       "setCurrentView          Set Current View\n",
       "setDefaultSandbox       setDefaultSandbox\n",
       "setDefaultSandboxPath   setDefaultSandboxPath\n",
       "setEdgeColorBypass      Set Edge Color Bypass\n",
       "setEdgeColorDefault     Set Edge Color Default\n",
       "setEdgeColorMapping     Set Edge Color Mapping\n",
       "setEdgeFontFaceBypass   Set Edge Font Face Bypass\n",
       "setEdgeFontFaceDefault\n",
       "                        Set Edge Font Face Default\n",
       "setEdgeFontFaceMapping\n",
       "                        Set Edge Font Face Mapping\n",
       "setEdgeFontSizeBypass   Set Edge Font Size Bypass\n",
       "setEdgeFontSizeDefault\n",
       "                        Set Edge Font Size Default\n",
       "setEdgeFontSizeMapping\n",
       "                        Set Edge Font Size Mapping\n",
       "setEdgeLabelBypass      Set Edge Label Bypass\n",
       "setEdgeLabelColorBypass\n",
       "                        Set Edge Label Color Bypass\n",
       "setEdgeLabelColorDefault\n",
       "                        Set Edge Label Color Default\n",
       "setEdgeLabelColorMapping\n",
       "                        Set Edge Label Color Mapping\n",
       "setEdgeLabelDefault     Set Edge Label Default\n",
       "setEdgeLabelMapping     Set Edge Label Mapping\n",
       "setEdgeLabelOpacityBypass\n",
       "                        Set Edge Label Opacity Bypass\n",
       "setEdgeLabelOpacityDefault\n",
       "                        Set Edge Label Opacity Default\n",
       "setEdgeLabelOpacityMapping\n",
       "                        Set Edge Label Opacity Mapping\n",
       "setEdgeLineStyleBypass\n",
       "                        Set Edge Line Style Bypass\n",
       "setEdgeLineStyleDefault\n",
       "                        Set Edge Line Style Default\n",
       "setEdgeLineStyleMapping\n",
       "                        Set Edge Line Style Mapping\n",
       "setEdgeLineWidthBypass\n",
       "                        Set Edge Line Width Bypass\n",
       "setEdgeLineWidthDefault\n",
       "                        Set Edge Line Width Default\n",
       "setEdgeLineWidthMapping\n",
       "                        Set Edge Line Width Mapping\n",
       "setEdgeOpacityBypass    Set Edge Opacity Bypass\n",
       "setEdgeOpacityDefault   Set Edge Opacity Default\n",
       "setEdgeOpacityMapping   Set Edge Opacity Mapping\n",
       "setEdgePropertyBypass   Set Edge Property Bypass\n",
       "setEdgeSelectionColorDefault\n",
       "                        Set Edge Selection Color Default\n",
       "setEdgeSourceArrowColorBypass\n",
       "                        Set Edge Source Arrow Color Bypass\n",
       "setEdgeSourceArrowColorDefault\n",
       "                        Set Edge Source Arrow Color Default\n",
       "setEdgeSourceArrowColorMapping\n",
       "                        Set Edge Source Arrow Color Mapping\n",
       "setEdgeSourceArrowMapping\n",
       "                        Set Edge Source Arrow Mapping\n",
       "setEdgeSourceArrowShapeBypass\n",
       "                        Set Edge Source Arrow Shape Bypass\n",
       "setEdgeSourceArrowShapeDefault\n",
       "                        Set Edge Source Arrow Shape Default\n",
       "setEdgeSourceArrowShapeMapping\n",
       "                        Set Edge Source Arrow Shape Mapping\n",
       "setEdgeTargetArrowColorBypass\n",
       "                        Set Edge Target Arrow Color Bypass\n",
       "setEdgeTargetArrowColorDefault\n",
       "                        Set Edge Target Arrow Color Default\n",
       "setEdgeTargetArrowColorMapping\n",
       "                        Set Edge Target Arrow Color Mapping\n",
       "setEdgeTargetArrowMapping\n",
       "                        Set Edge Target Arrow Mapping\n",
       "setEdgeTargetArrowShapeBypass\n",
       "                        Set Edge Target Arrow Shape Bypass\n",
       "setEdgeTargetArrowShapeDefault\n",
       "                        Set Edge Target Arrow Shape Default\n",
       "setEdgeTargetArrowShapeMapping\n",
       "                        Set Edge Target Arrow Shape Mapping\n",
       "setEdgeTooltipBypass    Set Edge Tooltip Bypass\n",
       "setEdgeTooltipDefault   Set Edge Tooltip Default\n",
       "setEdgeTooltipMapping   Set Edge Tooltip Mapping\n",
       "setLayoutProperties     Set Layout Properties\n",
       "setModelPropagationSecs\n",
       "                        Set Model Propagation Delay\n",
       "setNetworkCenterBypass\n",
       "                        Set Network Center Bypass\n",
       "setNetworkPropertyBypass\n",
       "                        Set Network Property Bypass\n",
       "setNetworkZoomBypass    Set Network Zoom Bypass\n",
       "setNodeBorderColorBypass\n",
       "                        Set Node Border Color Bypass\n",
       "setNodeBorderColorDefault\n",
       "                        Set Node Border Color Default\n",
       "setNodeBorderColorMapping\n",
       "                        Set Node Border Color Mapping\n",
       "setNodeBorderOpacityBypass\n",
       "                        Set Node Border Opacity Bypass\n",
       "setNodeBorderOpacityDefault\n",
       "                        Set Node Border Opacity Default\n",
       "setNodeBorderOpacityMapping\n",
       "                        Set Node Border Opacity Mapping\n",
       "setNodeBorderWidthBypass\n",
       "                        Set Node Border Width Bypass\n",
       "setNodeBorderWidthDefault\n",
       "                        Set Node Border Width Default\n",
       "setNodeBorderWidthMapping\n",
       "                        Set Node Border Width Mapping\n",
       "setNodeColorBypass      Set Node Color Bypass\n",
       "setNodeColorDefault     Set Node Color Default\n",
       "setNodeColorMapping     Set Node Color Mapping\n",
       "setNodeComboOpacityMapping\n",
       "                        Set Node Combo Opacity Mapping\n",
       "setNodeCustomBarChart   Set Node Custom Bar Chart\n",
       "setNodeCustomBoxChart   Set Node Custom Box Chart\n",
       "setNodeCustomHeatMapChart\n",
       "                        Set Node Custom HeatMap Chart\n",
       "setNodeCustomLinearGradient\n",
       "                        Set Node Custom Linear Gradient\n",
       "setNodeCustomLineChart\n",
       "                        Set Node Custom Line Chart\n",
       "setNodeCustomPieChart   Set Node Custom Pie Chart\n",
       "setNodeCustomPosition   Set Node Custom Position\n",
       "setNodeCustomRadialGradient\n",
       "                        Set Node Custom Radial Gradient\n",
       "setNodeCustomRingChart\n",
       "                        Set Node Custom Ring Chart\n",
       "setNodeFillOpacityBypass\n",
       "                        Set Node Fill Opacity Bypass\n",
       "setNodeFillOpacityDefault\n",
       "                        Set Node Fill Opacity Default\n",
       "setNodeFillOpacityMapping\n",
       "                        Set Node Fill Opacity Mapping\n",
       "setNodeFontFaceBypass   Set Node Font Face Bypass\n",
       "setNodeFontFaceDefault\n",
       "                        Set Node Font Face Default\n",
       "setNodeFontFaceMapping\n",
       "                        Set Node Font Face Mapping\n",
       "setNodeFontSizeBypass   Set Node Font Size Bypass\n",
       "setNodeFontSizeDefault\n",
       "                        Set Node Font Size Default\n",
       "setNodeFontSizeMapping\n",
       "                        Set Node Font Size Mapping\n",
       "setNodeHeightBypass     Set Node Height Bypass\n",
       "setNodeHeightDefault    Set Node Height Default\n",
       "setNodeHeightMapping    Set Node Height Mapping\n",
       "setNodeLabelBypass      Set Node Label Bypass\n",
       "setNodeLabelColorBypass\n",
       "                        Set Node Label Color Bypass\n",
       "setNodeLabelColorDefault\n",
       "                        Set Node Label Color Default\n",
       "setNodeLabelColorMapping\n",
       "                        Set Node Label Color Mapping\n",
       "setNodeLabelDefault     Set Node Label Default\n",
       "setNodeLabelMapping     Set Node Label Mapping\n",
       "setNodeLabelOpacityBypass\n",
       "                        Set Node Label Opacity Bypass\n",
       "setNodeLabelOpacityDefault\n",
       "                        Set Node Label Opacity Default\n",
       "setNodeLabelOpacityMapping\n",
       "                        Set Node Label Opacity Mapping\n",
       "setNodeOpacityBypass    Set Node Opacity Bypass\n",
       "setNodePropertyBypass   Set Node Property Bypass\n",
       "setNodeSelectionColorDefault\n",
       "                        Set Node Selection Color Default\n",
       "setNodeShapeBypass      Set Node Shape Bypass\n",
       "setNodeShapeDefault     Set Node Shape Default\n",
       "setNodeShapeMapping     Set Node Shape Mapping\n",
       "setNodeSizeBypass       Set Node Size Bypass\n",
       "setNodeSizeDefault      Set Node Size Default\n",
       "setNodeSizeMapping      Set Node Size Mapping\n",
       "setNodeTooltipBypass    Set Node Tooltip Bypass\n",
       "setNodeTooltipDefault   Set Node Tooltip Default\n",
       "setNodeTooltipMapping   Set Node Tooltip Mapping\n",
       "setNodeWidthBypass      Set Node Width Bypass\n",
       "setNodeWidthDefault     Set Node Width Default\n",
       "setNodeWidthMapping     Set Node Width Mapping\n",
       "setNotebookIsRunning    setNotebookIsRunning\n",
       "setSandboxReinitialize\n",
       "                        setSandboxReinitialize\n",
       "setStyleDependencies    Set Style Dependencies\n",
       "setVisualPropertyDefault\n",
       "                        Set Visual Property Default\n",
       "setVisualStyle          Set Visual Style\n",
       "spoofResponse-class     spoofResponse\n",
       "syncNodeCustomGraphicsSize\n",
       "                        Sync Node Custom Graphics Size\n",
       "toggleGraphicsDetails   Toggle Graphics Details\n",
       "ungroupAnnotation       Ungroup Annotation Group\n",
       "unhideAll               Unhide All\n",
       "unhideEdges             Unhide Edges\n",
       "unhideNodes             Unhide Nodes\n",
       "uninstallApp            Uninstall App\n",
       "UpdateAnnotationBoundedText\n",
       "                        Update Bounded Text Annotation\n",
       "updateAnnotationImage   Update Image Annotation\n",
       "updateAnnotationShape   Update Shape Annotation\n",
       "updateAnnotationText    Update Text Annotation\n",
       "updateApp               Update App\n",
       "updateGroupAnnotation   Update Group Annotation\n",
       "updateNetworkInNDEx     Update Network In NDEx\n",
       "updateStyleDefaults     Updates the default values of visual properties\n",
       "                        in a style\n",
       "updateStyleMapping      Updates a visual property mapping in a style"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "help(package=RCy3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05cfc5c2-d670-451e-952a-b8d5234c8654",
   "metadata": {},
   "source": [
    "### Help on specific cytoscape command\n",
    "To get information about an individual command from the R environment you can also use the commandsHelp function. Simply specify what command you would like to get information on by adding its name to the command. For example “commandsHelp(\"help string”)“"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e2e08da4-7827-47d3-8eb1-16e0db4bdf59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'analyzer'</li><li>'annotation'</li><li>'apps'</li><li>'command'</li><li>'cybrowser'</li><li>'cychart'</li><li>'diffusion'</li><li>'edge'</li><li>'filetransfer'</li><li>'filter'</li><li>'group'</li><li>'idmapper'</li><li>'layout'</li><li>'network'</li><li>'node'</li><li>'session'</li><li>'table'</li><li>'view'</li><li>'vizmap'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'analyzer'\n",
       "\\item 'annotation'\n",
       "\\item 'apps'\n",
       "\\item 'command'\n",
       "\\item 'cybrowser'\n",
       "\\item 'cychart'\n",
       "\\item 'diffusion'\n",
       "\\item 'edge'\n",
       "\\item 'filetransfer'\n",
       "\\item 'filter'\n",
       "\\item 'group'\n",
       "\\item 'idmapper'\n",
       "\\item 'layout'\n",
       "\\item 'network'\n",
       "\\item 'node'\n",
       "\\item 'session'\n",
       "\\item 'table'\n",
       "\\item 'view'\n",
       "\\item 'vizmap'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'analyzer'\n",
       "2. 'annotation'\n",
       "3. 'apps'\n",
       "4. 'command'\n",
       "5. 'cybrowser'\n",
       "6. 'cychart'\n",
       "7. 'diffusion'\n",
       "8. 'edge'\n",
       "9. 'filetransfer'\n",
       "10. 'filter'\n",
       "11. 'group'\n",
       "12. 'idmapper'\n",
       "13. 'layout'\n",
       "14. 'network'\n",
       "15. 'node'\n",
       "16. 'session'\n",
       "17. 'table'\n",
       "18. 'view'\n",
       "19. 'vizmap'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"analyzer\"     \"annotation\"   \"apps\"         \"command\"      \"cybrowser\"   \n",
       " [6] \"cychart\"      \"diffusion\"    \"edge\"         \"filetransfer\" \"filter\"      \n",
       "[11] \"group\"        \"idmapper\"     \"layout\"       \"network\"      \"node\"        \n",
       "[16] \"session\"      \"table\"        \"view\"         \"vizmap\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "commandsHelp(\"help\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aa63bf1-ffc1-4ec8-bfd6-a9b3243db788",
   "metadata": {},
   "source": [
    "### Cytoscape Basics\n",
    "Create a Cytoscape network from some basic R objects\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e1333e6a-b65d-4b63-b137-ef769106a4a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes <- data.frame(id=c(\"node 0\",\"node 1\",\"node 2\",\"node 3\"),\n",
    "    group=c(\"A\",\"A\",\"B\",\"B\"), # categorical strings\n",
    "    score=as.integer(c(20,10,15,5)), # integers\n",
    "    stringsAsFactors=FALSE)\n",
    "edges <- data.frame(source=c(\"node 0\",\"node 0\",\"node 0\",\"node 2\"),\n",
    "    target=c(\"node 1\",\"node 2\",\"node 3\",\"node 3\"),\n",
    "    interaction=c(\"inhibits\",\"interacts\",\"activates\",\"interacts\"),  # optional\n",
    "    weight=c(5.1,3.0,5.2,9.9), # numeric\n",
    "    stringsAsFactors=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eba8f3b9-6732-4fa8-b55f-ad047bfb42f4",
   "metadata": {},
   "source": [
    "#### Data frame used to create Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "08184e7b-e103-4655-9bf2-144066af9521",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>id</th><th scope=col>group</th><th scope=col>score</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>node 0</td><td>A</td><td>20</td></tr>\n",
       "\t<tr><td>node 1</td><td>A</td><td>10</td></tr>\n",
       "\t<tr><td>node 2</td><td>B</td><td>15</td></tr>\n",
       "\t<tr><td>node 3</td><td>B</td><td> 5</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 3\n",
       "\\begin{tabular}{lll}\n",
       " id & group & score\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t node 0 & A & 20\\\\\n",
       "\t node 1 & A & 10\\\\\n",
       "\t node 2 & B & 15\\\\\n",
       "\t node 3 & B &  5\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 3\n",
       "\n",
       "| id &lt;chr&gt; | group &lt;chr&gt; | score &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| node 0 | A | 20 |\n",
       "| node 1 | A | 10 |\n",
       "| node 2 | B | 15 |\n",
       "| node 3 | B |  5 |\n",
       "\n"
      ],
      "text/plain": [
       "  id     group score\n",
       "1 node 0 A     20   \n",
       "2 node 1 A     10   \n",
       "3 node 2 B     15   \n",
       "4 node 3 B      5   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d8ab7d74-4ea5-495b-a132-056defaad65c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 4 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>source</th><th scope=col>target</th><th scope=col>interaction</th><th scope=col>weight</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>node 0</td><td>node 1</td><td>inhibits </td><td>5.1</td></tr>\n",
       "\t<tr><td>node 0</td><td>node 2</td><td>interacts</td><td>3.0</td></tr>\n",
       "\t<tr><td>node 0</td><td>node 3</td><td>activates</td><td>5.2</td></tr>\n",
       "\t<tr><td>node 2</td><td>node 3</td><td>interacts</td><td>9.9</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 4 × 4\n",
       "\\begin{tabular}{llll}\n",
       " source & target & interaction & weight\\\\\n",
       " <chr> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t node 0 & node 1 & inhibits  & 5.1\\\\\n",
       "\t node 0 & node 2 & interacts & 3.0\\\\\n",
       "\t node 0 & node 3 & activates & 5.2\\\\\n",
       "\t node 2 & node 3 & interacts & 9.9\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 4 × 4\n",
       "\n",
       "| source &lt;chr&gt; | target &lt;chr&gt; | interaction &lt;chr&gt; | weight &lt;dbl&gt; |\n",
       "|---|---|---|---|\n",
       "| node 0 | node 1 | inhibits  | 5.1 |\n",
       "| node 0 | node 2 | interacts | 3.0 |\n",
       "| node 0 | node 3 | activates | 5.2 |\n",
       "| node 2 | node 3 | interacts | 9.9 |\n",
       "\n"
      ],
      "text/plain": [
       "  source target interaction weight\n",
       "1 node 0 node 1 inhibits    5.1   \n",
       "2 node 0 node 2 interacts   3.0   \n",
       "3 node 0 node 3 activates   5.2   \n",
       "4 node 2 node 3 interacts   9.9   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "edges"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9e9717d-d4d9-4fb1-b8de-e474ad4c57dd",
   "metadata": {},
   "source": [
    "#### Create Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "80788e7a-0120-4d38-88ec-22e1bb8f5d38",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading data...\n",
      "\n",
      "\n",
      "Applying default style...\n",
      "\n",
      "\n",
      "Applying preferred layout...\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>networkSUID:</strong> 124"
      ],
      "text/latex": [
       "\\textbf{networkSUID:} 124"
      ],
      "text/markdown": [
       "**networkSUID:** 124"
      ],
      "text/plain": [
       "networkSUID \n",
       "        124 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "createNetworkFromDataFrames(nodes,edges, title=\"my first network\", \n",
    "    collection=\"DataFrame Example\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bfc6931-b396-4270-b327-45d66ba11086",
   "metadata": {},
   "source": [
    "#### Export an image of the network\n",
    "Remember. All networks we make are created in Cytoscape so get an image of the resulting network and include it in your current analysis if desired."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7e3523b1-9467-4677-a04c-2c04d104216b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>file:</strong> 'C:\\\\Users\\\\kozo2\\\\CytoscapeConfiguration\\\\filetransfer\\\\default_sandbox\\\\my_first_network.png'"
      ],
      "text/latex": [
       "\\textbf{file:} 'C:\\textbackslash{}\\textbackslash{}Users\\textbackslash{}\\textbackslash{}kozo2\\textbackslash{}\\textbackslash{}CytoscapeConfiguration\\textbackslash{}\\textbackslash{}filetransfer\\textbackslash{}\\textbackslash{}default\\_sandbox\\textbackslash{}\\textbackslash{}my\\_first\\_network.png'"
      ],
      "text/markdown": [
       "**file:** 'C:\\\\Users\\\\kozo2\\\\CytoscapeConfiguration\\\\filetransfer\\\\default_sandbox\\\\my_first_network.png'"
      ],
      "text/plain": [
       "                                                                                           file \n",
       "\"C:\\\\Users\\\\kozo2\\\\CytoscapeConfiguration\\\\filetransfer\\\\default_sandbox\\\\my_first_network.png\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "exportImage(\"my_first_network\", type = \"png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be64ad3f-ab2e-41a1-af70-17b0dcc077b0",
   "metadata": {},
   "source": [
    "## Guided Walkthrough\n",
    "\n",
    "[Cytoscape Rmd Notebooks](https://cytoscape.org/cytoscape-automation/for-scripters/R/notebooks/)\n",
    "\n",
    "- Overview of RCy3\n",
    "- Cancer networks and data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea3e3398-4130-419f-bb9d-f6a7ec878d9f",
   "metadata": {},
   "source": [
    "## Example Use Cases\n",
    "\n",
    "1. **Omics data** - I have a ----------- fill in the blank (microarray, RNASeq, Proteomics, ATACseq, MicroRNA, GWAS …) dataset. I have normalized and scored my data. How do I overlay my data on existing interaction data?\n",
    "2. **Coexpression data** -I have a dataset that represents relationships. How do I represent it as a network.\n",
    "3. **Omics data** - I have a -----------fill in the blank (microarray, RNASeq, Proteomics, ATACseq, MicroRNA, GWAS …) dataset. I have normalized and scored my data. I have run my data through a functional enrichment tool and now have a set of enriched terms associated with my dataset. How do I represent my functional enrichments as a network?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0983ea3-0aa1-46e6-b78d-6dc57bf15e99",
   "metadata": {},
   "source": [
    "### Example Data Set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39a1fdee-5d6f-4833-aabf-14f19384b242",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
